src/connection/connection_support.c

```diff
@@ -645,6 +645,164 @@ css_net_recv (SOCKET fd, char *buffer, int *maxlen, int timeout)
+	  return ERROR_WHEN_READING_SIZE;
+	}
+      if (nbytes_read_len != sizeof (int))
+	{
+#ifdef CUBRID_DEBUG
+	  er_log_debug (ARG_FILE_LINE, "css_net_recv: returning ERROR_WHEN_READING_SIZE bytes %d \n", nbytes_read_len);
+#endif
+	  return ERROR_WHEN_READING_SIZE;
+	}
+      else
+	{
+	  break;
+	}
+    }
+
+  templen = ntohl (templen);
+  int length_to_read = 0;
+  if (templen > *maxlen)
+    {
+      length_to_read = *maxlen;
+    }
+  else
+    {
+      length_to_read = templen;
+    }
+
+  /* read data */
+  const int nbytes_read_data = css_readn (fd, buffer, length_to_read, timeout);
+  if (nbytes_read_data < length_to_read)
+    {
+#ifdef CUBRID_DEBUG
+      er_log_debug (ARG_FILE_LINE, "css_net_recv_allow_truncated: returning ERROR_ON_READ bytes %d\n",
+		    nbytes_read_data);
+#endif
+      return ERROR_ON_READ;
+    }
+
+  /*
+   * This is possible if the data buffer provided by the client is smaller
+   * than the number of bytes sent by the server
+   */
+
+  if ((nbytes_read_data > 0) && (templen > nbytes_read_data))
+    {
+      // report remaining length
+      *remlen = templen - nbytes_read_data;
+      return RECORD_TRUNCATED;
+    }
+
+  if (nbytes_read_data != templen)
+    {
+#ifdef CUBRID_DEBUG
+      er_log_debug (ARG_FILE_LINE, "css_net_recv: returning READ_LENGTH_MISMATCH bytes %d\n", nbytes_read_data);
+#endif
+      return READ_LENGTH_MISMATCH;
+    }
+
+  *maxlen = nbytes_read_data;
+  return NO_ERRORS;
+}
+
+/*
+ * css_net_recv_remainder - read the remainder number of bytes from the supplied socket; the function
+ *                expects to receive an appropriately sized buffer and to be able to read exactly
+ *                that many bytes; this function must only be called if a previous call to
+ *                css_net_recv_allow_truncated returned the error code RECORD_TRUNCATED
+ *
+ *   return: 0 if success, or error code
+ *   fd(in): socket descriptor
+ *   buffer(out): buffer for data be read
+ *   maxlen(in/out):
+ *          in - length of buffer, must be exactly the same as the number of remaining bytes to be read;
+ *          out - count of bytes that were read; exactly the same value as input
+ *   timeout(in): timeout value in milli-second
+ */
+css_error_code
+css_net_recv_remainder (SOCKET fd, char *buffer, int *maxlen, int timeout)
+{
+  /* read remainder data */
+  const int nbytes_read_data = css_readn (fd, buffer, *maxlen, timeout);
+  if (nbytes_read_data < *maxlen)
+    {
+      // expecting to be able to read exactly as expected
+#ifdef CUBRID_DEBUG
+      er_log_debug (ARG_FILE_LINE, "css_net_recv_remainder: ERROR_ON_READ bytes %d\n", nbytes_read_data);
+#endif
+      return ERROR_ON_READ;
+    }
+
+  if (nbytes_read_data != *maxlen)
+    {
+      // this should never occur
+#ifdef CUBRID_DEBUG
+      er_log_debug (ARG_FILE_LINE, "css_net_recv_remainder: READ_LENGTH_MISMATCH bytes %d\n", nbytes_read_data);
+#endif
+      assert (false);
+      return READ_LENGTH_MISMATCH;
+    }
```

Comment:
The condition 'nbytes_read_data != *maxlen' is actually 'nbytes_read_data > *maxlen' because 'nbytes_read_data < *maxlen' is already checked in the previous if condition. I don't think it can happen. What is it for?